File: sample.pl
#!/usr/bin/env perl
##!/usr/bin/perl
#mark a#
# Sample script for perl users to use quaternary checksum, redundancy and their combination as well as quaternary Hamming code.

use v5.21.1;
use utf8;
use strict;
use warnings;
# use YAML;   # say Dump
# use POSIX qw/ ceil floor /;
use Data::Dump qw/ dump /;  #dump @AoA;
# use lib $ENV{HOME} . '/perl/lib/';
use lib './lib/';
use QEDC qw/ 
  encode_q_cs_v1 
  encode_q_cs_v2 
  is_intact_q_cs_v1 
  is_intact_q_cs_v2 
  decode_q_cs_v1 
  decode_q_cs_v2 
  encode_q_redun_v1 
  encode_q_redun_v2 
  is_intact_q_redun_v1 
  decode_q_redun_v1 
  decode_q_redun_v2 
  encode_comb_q_cs_redun 
  decode_comb_q_cs_redun 
  encode_q_hamming 
  decode_q_hamming 
  ref_to_RAR 
  /;
#mark b#
# naming convention: 1. A: array; 2. H: hash; 3. R: reference; 4. Sr: subroutine; 5. crt: correct; 6. err: error; 7. RAR: references to array of references; 8. 

say '================================================';
say "Encoding algorithm of quaternary checksum";
my $a_in = "0123";           # change this as necessary
print '$a_in   '; say $a_in;
my ($cs, $a_out, ) = & encode_q_cs_v1 ($a_in, );
print '($cs, $a_out, )  '; map {print "|$_"; } ($cs, $a_out, ); say '|';

say '================================================';
say "Detect the occurence of errors using quaternary checksum";
# my $a_in = "30123";          # change this as necessary
my $a_in = "20123";          # change this as necessary
my $is_correct = & is_intact_q_cs_v1 ($a_in, );
print '$is_correct   '; say $is_correct;

say '================================================';
say "Decoding algorithm of quaternary checksum";
my $a_in = "30123";          # change this as necessary
my $q = 2;                   # change this as necessary
my $a_crt = & decode_q_cs_v1 ($a_in, $q, );
print 'dump ($a_in, $q, $a_crt, )   '; say dump ($a_in, $q, $a_crt, );

say '================================================';
say "Encoding algorithm of quaternary redundancy";
my $a_R = ["01223", "123", "033201", ];   # change this as necessary
print 'dump $a_R   '; say dump $a_R;
my ($redun, $c_R, ) = & encode_q_redun_v1 ($a_R, );
print '$redun   '; say $redun;
print 'dump $c_R   '; say dump $c_R;

say '================================================';
say "Detect the occurence of errors using quaternary redundancy";
# change this as necessary
my $a_R = ["01223", "123", "033201", "120031"];   # intact
# my $a_R = ["01223", "023", "033201", "120031"]; # erroneous
# my $a_R = ["01223", "12", "033201", "120031"]; # erroneous
my $is_correct = & is_intact_q_redun_v1 ($a_R, );
print '$is_correct   '; say $is_correct;

say '================================================';
say "Decoding algorithm of quaternary redundancy";
# change this as necessary
# my $a_R = ["01223", "123", "033201", "120031"];   # intact
# my $a_R = ["01223", "023", "033201", "120031"]; # erroneous
my $a_R = ["01223", "12", "033201", "120031"]; # erroneous
print 'dump $a_R   '; say dump $a_R;
my $c_R = & decode_q_redun_v1 ($a_R, 2, 3 );
print 'dump $c_R   '; say dump $c_R;

say '================================================';
say "Encoding algorithm of the combination of quaternary checksum and redundancy";
say 'Example 1, for sequences of possibly different length';
# change this as necessary
my $a_R = ["01223", "123", "033201"];   
print 'dump $a_R   '; say dump $a_R;
my $c_R = & encode_comb_q_cs_redun ($a_R, );
print 'dump $c_R   '; say dump $c_R;

say '================================================';
say 'Decoding algorithm of the combination of quaternary checksum and redundancy';
say 'Example 1, for sequences of possibly different length';
# change this as necessary
# my $c_R = ["001223", "2123", "1033201", "3120031"];   # intact
my $c_R = ["001", "2123", "1033201", "3120031"];   # erroneous
my $out_R = & decode_comb_q_cs_redun ($c_R, 5, );
print 'dump $out_R   '; say dump $out_R;

say '================================================';
say "Encoding algorithm of the combination of quaternary checksum and redundancy";
say 'Example 2, for sequences of the same length';
# change this as necessary
# my $a_R = ["01223", "12300", "03320"];  
my $a_R = ["0122310132", "1230022132", "0332013213"];  
print 'dump $a_R   '; say dump $a_R;
my $c_R = & encode_comb_q_cs_redun ($a_R, );
print 'dump $c_R   '; say dump $c_R;

say '================================================';
say "Decoding algorithm of the combination of quaternary checksum and redundancy";
say 'Example 2, for sequences of the same length';
# change this as necessary
# my $c_R = ["001223", 212300, "003320", 212003];   # intact
my $c_R = ["001", 212300, "003320", 212003];   # erroneous
my $out_R = & decode_comb_q_cs_redun ($c_R, 5, );
print 'dump $out_R   '; say dump $out_R;

say '================================================';
say "Encoding algorithm of quaternary Hamming code";
# change this as necessary
# my $a_in = "0123" . "2311" . "010";   # $c : 320012302311010
# my $a_in = "3102302031020310201";       # $c : 313210223020310120310201
my $a_in = "0122310132";       # $c : 13031222310132
my $c = & encode_q_hamming ($a_in, );
print 'dump ($a_in, $c, )   '; say dump ($a_in, $c, );

say '================================================';
say "Decoding algorithm of quaternary Hamming code";
# change this as necessary
# my $c = "310012302311010";              # $a : 01232311010
# my $c = "313210223020310120310201";   # $a : 3102302031020310201
# my $c = "313210223020310120310211";   # $a : 3102302031020310201
my $c = "13031222310132";   # $a : 0122310132
my $a = & decode_q_hamming ($c, );    
print 'dump ($c, $a, )   '; say dump ($c, $a, );

#mark z#

File: sample.py
#!/usr/bin/env python3
##!/usr/bin/env python2
# Sample script for python users to use quaternary checksum, redundancy and their combination as well as quaternary Hamming code. Works for both python2 and python3.

import subprocess
import re
#mark a#

#mark b#
## type                   | input data
# encode_q_cs             | a string
# is_intact_q_cs          | a string
# decode_q_cs             | string, string
# encode_q_redun          | array of strings 
# is_intact_q_redun       | array of strings
# decode_q_redun          | array of strings, string, string
# encode_comb_q_cs_redun  | array of strings
# decode_comb_q_cs_redun  | array of strings, string 
# encode_q_hamming        | a string
# decode_q_hamming        | a string

print('================================================')
print('Encoding algorithm of quaternary checksum')
if 'result' in locals():
  del result
if 'string' in locals(): 
  del string 
# Encode the input data with quaternary checksum prepended
type = "encode_q_cs"    
data = "200131233101231"    # change this as necessary
print( 'type   ', type)
print( 'data   ', data)
cmd = ["./lib/QEDC_py.pl", type, data]
pipe = subprocess.Popen(cmd, stdout = subprocess.PIPE)
result = pipe.stdout.read()
encoded_data = result.decode('UTF-8')   # 3200131233101231
print( 'encoded_data   ', encoded_data)   

print('================================================')
print('Detect the occurence of errors using quaternary checksum')
if 'result' in locals():
  del result
if 'string' in locals(): 
  del string 
# Check whether the encoded data is correct, of which the first number is quaternary checksum
# data : the encoded data
# 1 : intact
# 0 : erroneous
type = "is_intact_q_cs"    
data = "3200131233101231"    # change this as necessary
print( 'type   ', type)
print( 'data   ', data)
cmd = ["./lib/QEDC_py.pl", type, data]
pipe = subprocess.Popen(cmd, stdout = subprocess.PIPE)
result = pipe.stdout.read()
is_correct = result.decode('UTF-8')   
print( 'is_correct   ', is_correct)

print('================================================')
print('Decoding algorithm of quaternary checksum')
if 'result' in locals():
  del result
if 'string' in locals(): 
  del string 
# data : the encoded data, of which the first number is quaternary checksum and the (q+1)th number is erroneous (counting from 1), q > 1
type = "decode_q_cs"
# change this as necessary
data = "3210131233101231"   # the third number is erroneous
q = "2"   # the data type should be string
print( 'type   ', type)
print( 'data   ', data)
print( 'q   ', q)
cmd = ["./lib/QEDC_py.pl", type, data, q]
pipe = subprocess.Popen(cmd, stdout = subprocess.PIPE)
result = pipe.stdout.read()
string = result.decode('UTF-8')   
decoded_data = string
print( 'decoded_data   ', decoded_data)

print('================================================')
print('Encoding algorithm of quaternary redundancy')
if 'result' in locals():
  del result
if 'string' in locals(): 
  del string 
type = "encode_q_redun"   
data = ['0123', '22301', '3011032']   # change this as necessary
data = ' '.join(data)   # "0123 22301 3011032"
print( 'type   ', type)
print( 'data   ', data)
cmd = ["./lib/QEDC_py.pl", type, data]
pipe = subprocess.Popen(cmd, stdout = subprocess.PIPE)
result = pipe.stdout.read()
string = result.decode('UTF-8')
regex = r'(\d+)'
encoded_data = re.findall(regex, string)
print( 'encoded_data   ', encoded_data)
redun = encoded_data[-1]
print( 'redun   ', redun)

print('================================================')
print('Detect the occurence of errors using quaternary redundancy')
if 'result' in locals():
  del result
if 'string' in locals(): 
  del string 
# 1 : intact
# 0 : erroneous
type = "is_intact_q_redun"   
data = ['0123', '22301', '3011032', '1320132']    # intact, return "1"
# data = ['0123', '101', '3011032', '1320132']      # erroneous, return "0"
data = ' '.join(data)
print( 'type   ', type)
print( 'data   ', data)
cmd = ["./lib/QEDC_py.pl", type, data]
pipe = subprocess.Popen(cmd, stdout = subprocess.PIPE)
result = pipe.stdout.read()
is_correct = result.decode('UTF-8')
print( 'is_correct   ', is_correct)

print('================================================')
print('Decoding algorithm of quaternary redundancy')
if 'result' in locals():
  del result
if 'string' in locals(): 
  del string 
type = "decode_q_redun"   
# The correct encoded data is ['0123', '22301', '3011032', '1320132']
# After decoding : ['0123', '22301', '3011032']
data = ['0123', '101', '3011032', '1320132']      # change this as necessary
q = '2'
nq = '5'
data = ' '.join(data)     # "0123 101 3011032 1320132"
print( 'type   ', type)
print( 'data   ', data)
print( 'nq   ', nq)
cmd = ["./lib/QEDC_py.pl", type, data, q, nq]
pipe = subprocess.Popen(cmd, stdout = subprocess.PIPE)
result = pipe.stdout.read()
string = result.decode('UTF-8')
regex = r'(\d+)'
decoded_data = re.findall(regex, string)
print( 'decoded_data   ', decoded_data)

print('================================================')
print('Encoding algorithm of the combination of quaternary checksum and redundancy')
print('Example 1, for sequences of possibly different length')
if 'result' in locals():
  del result
if 'string' in locals(): 
  del string 
type = "encode_comb_q_cs_redun"   
data = ["01223", "123", "033201"]    # change this as necessary
data = ' '.join(data)   
print( 'type   ', type)
print( 'data   ', data)
cmd = ["./lib/QEDC_py.pl", type, data]
pipe = subprocess.Popen(cmd, stdout = subprocess.PIPE)
result = pipe.stdout.read()
string = result.decode('UTF-8')
regex = r'(\d+)'
encoded_data = re.findall(regex, string)
print( 'encoded_data   ', encoded_data)

print('================================================')
print('Decoding algorithm of the combination of quaternary checksum and redundancy')
print('Example 1, for sequences of possibly different length')
if 'result' in locals():
  del result
if 'string' in locals(): 
  del string 
type = "decode_comb_q_cs_redun"   
# change data as necessary
data = ["001223", "2123", "1033201", "3120031"];   # intact
data = ["001", "2123", "1033201", "3120031"];   # erroneous
data = ' '.join(data)
nq = "5"
print( 'type   ', type)
print( 'data   ', data)
print( 'nq   ', nq)
cmd = ["./lib/QEDC_py.pl", type, data, nq]
pipe = subprocess.Popen(cmd, stdout = subprocess.PIPE)
result = pipe.stdout.read()
string = result.decode('UTF-8')
print( 'string   ', string)
regex = r'(\d+)'
decoded_data = re.findall(regex, string)
print( 'decoded_data   ', decoded_data)

print('================================================')
print('Encoding algorithm of the combination of quaternary checksum and redundancy')
print('Example 2, for sequences of the same length ')
if 'result' in locals():
  del result
if 'string' in locals(): 
  del string 
type = "encode_comb_q_cs_redun"   
data = ["01223", "12300", "03320"]   # change this as necessary
data = ' '.join(data)   
print( 'type   ', type)
print( 'data   ', data)
cmd = ["./lib/QEDC_py.pl", type, data]
pipe = subprocess.Popen(cmd, stdout = subprocess.PIPE)
result = pipe.stdout.read()
string = result.decode('UTF-8')
regex = r'(\d+)'
encoded_data = re.findall(regex, string)
print( 'encoded_data   ', encoded_data)

print('================================================')
print('Decoding algorithm of the combination of quaternary checksum and redundancy')
print('Example 2, for sequences of the same length')
if 'result' in locals():
  del result
if 'string' in locals(): 
  del string 
type = "decode_comb_q_cs_redun"   
# change data as necessary
# data = ["001223", "212300", "003320", "212003"];   # intact
data = ["001", "212300", "003320", "212003"];   # erroneous
data = ' '.join(data)
nq = "5"
print( 'type   ', type)
print( 'data   ', data)
print( 'nq   ', nq)
cmd = ["./lib/QEDC_py.pl", type, data, nq]
pipe = subprocess.Popen(cmd, stdout = subprocess.PIPE)
result = pipe.stdout.read()
string = result.decode('UTF-8')
print( 'string   ', string)
regex = r'(\d+)'
decoded_data = re.findall(regex, string)
print( 'decoded_data   ', decoded_data)

print('================================================')
print('Encoding algorithm of quaternary Hamming code')
if 'result' in locals():
  del result
if 'string' in locals(): 
  del string 
type = "encode_q_hamming"   
data = "01232311010"   # change this as necessary
print( 'type   ', type)
print( 'data   ', data)
cmd = ["./lib/QEDC_py.pl", type, data]
pipe = subprocess.Popen(cmd, stdout = subprocess.PIPE)
result = pipe.stdout.read()
encoded_data = result.decode('UTF-8')
print( 'encoded_data   ', encoded_data)

print('================================================')
print('Decoding algorithm of quaternary Hamming code')
if 'result' in locals():
  del result
if 'string' in locals(): 
  del string 
type = "decode_q_hamming"   
data = "320012302311010"   # change this as necessary
print( 'type   ', type)
print( 'data   ', data)
cmd = ["./lib/QEDC_py.pl", type, data]
pipe = subprocess.Popen(cmd, stdout = subprocess.PIPE)
result = pipe.stdout.read()
decoded_data = result.decode('UTF-8')
print( 'decoded_data   ', decoded_data)
